src/array/cpu/rowwise_pick.h

```diff
@@ -310,101 +309,116 @@ COOMatrix CSRRowWisePerEtypePick(
         std::vector<IdxType> cols;
         std::vector<IdxType> idx;
 
-        std::vector<IdxType> et(len);
+        std::vector<IdxType> et;
         std::vector<IdxType> et_idx(len);
-        std::vector<IdxType> et_eid(len);
+        std::vector<IdxType> et_eid;
         std::iota(et_idx.begin(), et_idx.end(), 0);
-        for (int64_t j = 0; j < len; ++j) {
-          const IdxType homogenized_eid = eid ? eid[off + j] : off + j;
-          auto it = std::upper_bound(
-              eid2etype_offset.begin(), eid2etype_offset.end(),
-              homogenized_eid);
-          const IdxType heterogenized_etype = it - eid2etype_offset.begin() - 1;
-          const IdxType heterogenized_eid =
-              homogenized_eid - eid2etype_offset[heterogenized_etype];
-          et[j] = heterogenized_etype;
-          et_eid[j] = heterogenized_eid;
+        // heterogenized eid is needed when
+        // 1. neighbors of each node is not sorted.
+        // 2. probs_or_mask is not empty as it is indexed by heterogenized eid.
+        if (!rowwise_etype_sorted || has_probs) {
+          et.resize(len);
+          et_eid.resize(len);
```

Comment:
What if the condition does not hold (or does it *always* hold)?  Would `et` and `et_eid` be a size-0 vector and cause problems later on?

Reply:
It is by design, we use the `IsEmpty` to determine if `etype` and `heterogenized_eid` is calculated in advance in later code.

Reply:
I saw the `empty()` call in L349-351, but in L384-385, `et` and `et_eid` are still being directly accessed, no?

Reply:
In latest code, it will not be accessed in these 2 lines.